### ISSUE TITLE
stft output shape is not right

### ISSUE THREAD

**Issue by tom-huntington** [pytorch/pytorch#87584]
### 🐛 Describe the bug

When calculating the output shape using the convolution equation

```
channels_out = (in_channels-(kernel_size-1)-1)/stride+1
```

The kernel size should be the `win_length` but instead your implementation is using `n_fft`.

```
import torch

in_channels = 1000
stride = 100
kernel_size = 16
input = torch.zeros((in_channels,))

for i in range(4, 9):
    n_fft = 2**i
    output = torch.stft(input, n_fft=n_fft, hop_length=stride, win_length=kernel_size, center=False, return_complex=True)
    expected_channels_out = (in_channels-(kernel_size-1)-1)/stride+1
    actual_channels_out = (in_channels-(n_fft-1)-1)/stride+1
    print(f"{output.shape=}, {expected_channels_out=}, {actual_channels_out=}")
```

I guess this can be gotten around by padding the end of the input by `n_fft-win_length`.

### Versions

Collecting environment information...
PyTorch version: 1.11.0+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.1.0-1ubuntu1~20.04) 11.1.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.10 (default, Mar 15 2022, 12:22:08)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.10.43.3-microsoft-standard-WSL2-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to:
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 497.29
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.3
[pip3] torch==1.11.0+cu113
[pip3] torchaudio==0.11.0+cu113
[pip3] torchvision==0.12.0+cu113
[conda] Could not collect

cc @mruberry @peterbell10

**Comment by peterbell10** [comment]
When the STFT window is smaller than `n_fft`, it gets center padded to size `n_fft` e.g. `win_length=2`, `n_fft=4` would be the same as passing `window=tensor([0, 1, 1, 0])`. In that sense, the kernel really is always of size `n_fft`. This behavior is inherited from `librosa.stft` which `torch.stft` was modelled after.

To match your assumptions, I think you need to manually pass a right-padded window e.g. `tensor([1, 1, 0, 0])` as well as padding the input sequence as you mentioned.

**Comment by tom-huntington** [comment]
> To match your assumptions, I think you need to manually pass a right-padded window e.g. `tensor([1, 1, 0, 0])` as well as padding the input sequence as you mentioned.

Will do thanks! I should have read the docs more throughly

> window can be a 1-D tensor of size win_length, e.g., from [torch.hann_window()](https://pytorch.org/docs/stable/generated/torch.hann_window.html#torch.hann_window). If window is None (default), it is treated as if having 11 everywhere in the window. **If win_length < n_fft**, window will be padded on both sides to length n_fft before being applied.

Obviously this API is non-sense, because zero padding the fft is just increasing the resolution of the fft (i.e. interpolates the fft frequencies) and has no relation to the window (i.e. time resolution). https://dsp.stackexchange.com/questions/37927/what-happens-when-n-increases-in-n-point-dft/37931#37931
